Makefile

```diff
@@ -2,127 +2,157 @@ MPI ?= 1
 ADDUS ?= 1
 USREXIT ?= 0
 NBC ?= 0
-LIBNAME ?= gs
 BLAS ?= 0
-DEBUG ?= 0
+
 CFLAGS ?= -O2
 FFLAGS ?= -O2
+ARFLAGS ?= cr
+
+LIBNAME ?= gs
 CPREFIX ?= gslib_
 FPREFIX ?= fgslib_
 
-SRCROOT=.
-TESTDIR=$(SRCROOT)/tests
-FTESTDIR=$(TESTDIR)/fortran
-SRCDIR=$(SRCROOT)/src
-INCDIR=$(SRCROOT)/src
-LIBDIR=$(SRCROOT)/lib
+STATIC ?= 1
+SHARED ?= 0
+
+SRCROOT = .
+TESTDIR = $(SRCROOT)/tests
+FTESTDIR = $(TESTDIR)/fortran
+SRCDIR = $(SRCROOT)/src
+INCDIR = $(SRCROOT)/src
+LIBDIR = $(SRCROOT)/lib
+
+DARWIN := $(filter Darwin,$(shell uname -s))
+SO_EXT := $(if $(DARWIN),dylib,so)
 
 ifneq (,$(strip $(DESTDIR)))
-INSTALL_ROOT = $(DESTDIR)
+  INSTALL_ROOT = $(DESTDIR)
 else
-INSTALL_ROOT = $(SRCROOT)/build
+  INSTALL_ROOT = $(SRCROOT)/build
+endif
+
+ifneq (0,$(SHARED))
+  ifneq (0,$(STATIC))
+    $(warning Cannot build with both STATIC=1 and SHARED=1, setting SHARED=0)
+    override SHARED = 0
+  endif
+endif
+
+ifneq (0,$(SHARED))
+  ifeq ($(filter -fPIC,$(CFLAGS)),)
+    override CFLAGS += -fPIC
+  endif
+  ifneq ($(DARWIN),)
```

Comment:
Why is rpath only specified on Darwin?

Reply:
I think `-install_name` is only a macOS linker flag (https://gcc.gnu.org/onlinedocs/gcc-3.3.6/gcc/Darwin-Options.html). This is borrowed from https://github.com/CEED/libCEED/blob/main/Makefile, and was necessary to get the dynamic library to load correctly on macOS. On Linux it worked fine without any changes.